packages/cli/openapi-parser/src/openapi/v3/AbstractOpenAPIV3ParserContext.ts

```diff
@@ -46,38 +46,86 @@ export abstract class AbstractOpenAPIV3ParserContext implements SchemaParserCont
-            if (resolvedProperty == null) {
-                throw new Error(`${schema.$ref} is undefiened. Property does not exist on object.`);
-            } else if (isReferenceObject(resolvedProperty)) {
-                resolvedSchema = this.resolveSchemaReference(resolvedProperty);
-            } else {
-                resolvedSchema = resolvedProperty;
+            if (splitSchemaKey[0] == null) {
+                throw new Error(`${schema.$ref} is undefined`);
             }
-        }
+            resolvedSchema = this.document.components.schemas[splitSchemaKey[0]];
+            if (resolvedSchema == null) {
+                throw new Error(`${splitSchemaKey[0]} is undefined`);
+            }
+            if (isReferenceObject(resolvedSchema)) {
+                resolvedSchema = this.resolveSchemaReference(resolvedSchema);
+            }
+
+            if (splitSchemaKey[1] === "properties" && splitSchemaKey[2] != null) {
+                const resolvedProperty = resolvedSchema.properties?.[splitSchemaKey[2]];
+                if (resolvedProperty == null) {
+                    throw new Error(`${schema.$ref} is undefiened. Property does not exist on object.`);
+                } else if (isReferenceObject(resolvedProperty)) {
+                    resolvedSchema = this.resolveSchemaReference(resolvedProperty);
+                } else {
+                    resolvedSchema = resolvedProperty;
+                }
+            }
+
+            return resolvedSchema;
+        } else {
+            schemaKey = schema.$ref.substring(2);
+            const parts = schemaKey.split("/");
+            let currentObject: any = this.document;
+
+            for (const part of parts) {
+                // Replace any escaped characters back to their original form
+                const key = part.replace(/~1/g, "/");
 
-        return resolvedSchema;
+                // Access the next nested property
+                currentObject = currentObject[key];
+            }
+
+            resolvedSchema = currentObject;
+
+            if (resolvedSchema == null) {
+                throw new Error(`${schemaKey} is undefined`);
+            }
+
+            if (isReferenceObject(resolvedSchema)) {
+                resolvedSchema = this.resolveSchemaReference(resolvedSchema);
+            }
+
+            if (parts[1] === "properties" && parts[2] != null) {
+                const resolvedProperty = resolvedSchema.properties?.[parts[2]];
+                if (resolvedProperty == null) {
+                    throw new Error(`${schema.$ref} is undefiened. Property does not exist on object.`);
+                } else if (isReferenceObject(resolvedProperty)) {
+                    resolvedSchema = this.resolveSchemaReference(resolvedProperty);
+                } else {
+                    resolvedSchema = resolvedProperty;
+                }
+            }
+
+            const schemaId = schema.$ref;
```

Comment:
The names assigned to schemas directly have an impact on the quality of the fern-generated SDK. We should be smarter about how we generate names: 

So if the ref points to a schema (`#/components/schemas/MyUser`) we should use the `MyUser`. If the ref points to a response (`#/components/responses/MyResponse/application~json/200/scheam` we should use `MyResponse`. If the ref points to a nested property (`#/components/schemas/MyUser/properties/ssn` we should make it `MyUserSsn`. 

I think it may be worthwhile to create a static helper function called `getSchemaIdFromRef` and unit test that as well.